#===============================================================
# Title: Spatial Transcriptomics Analysis from Nanostring GeoMx DSP data
# Author: Xiaoyue Deng, Mansour Faye, Anke Augspach (Corresponding Author)
# Contact: xiaoyue.deng@students.unibe.ch, anke.augspach@unibe.ch
# Date: 2024-09-12
# License: MIT
#===============================================================

#====== load libraries ======
library(tidyverse)
library(miscTools)
library(readxl)
library("org.Hs.eg.db")
library(devtools)  
library(remotes)
library(dplyr)
library(ggplot2)
library(ggrepel)
#library(gridExtra)
library(openxlsx)
library(clusterProfiler)
library(DESeq2)
library('enrichplot')
library("EnhancedVolcano")
library('pheatmap')

#======Loading data frames and generating required variables======
## load annotation and count datasets 
annot_data <- read.csv("File_Path")
count_data <- read.csv("File_Path", check.names = FALSE)
setwd("Your_WD")
target_properties <- read_excel(file_path,sheet = 2)
count_data <- read_excel(file_path,sheet = 3)
bio_probe <- read_excel(file_path, sheet = 4)
# =====Subset the results (Optional)=====
tumour_annot <- annot_data[annot_data$SegmentLabel == "tumour", ]
tme_annot <- annot_data[annot_data$SegmentLabel == "TME", ]
#annot_data <- tumour_annot
#annot_data <- tme_annot
annot_data <- annot_data %>% mutate("pAlignedReads"=AlignedReads/RawReads*100)
annot_data <- annot_data %>% mutate("ExpressionFilteringThreshold..Human.NGS.Whole.Transcriptome.Atlas.RNA_1.0."=
                                      ifelse("LOQ (Human NGS Whole Transcriptome Atlas RNA_1.0"<=2, 2,
                                             ifelse("LOQ (Human NGS Whole Transcriptome Atlas RNA_1.0">2, "LOQ (Human NGS Whole Transcriptome Atlas RNA_1.0", NA)))
## filter annot_data segments removing those segments with <60% pAlignedReads
annot_data.1 <- annot_data %>% filter("pAlignedReads">=60)
## select samples in count_data by those remaining in annot_data.1 following segment filtering by pAlignedReads
count_data.1 <- count_data %>% dplyr::select(TargetName, annot_data.1$SegmentDisplayName)
################# Filtering steps for segments and targets #################
### THIS CODE WAS GENERATED BY MOHAMED MANSOUR FAYE - PLEASE ACKNOWLEDGE ACCORDINGLY IN ANY DERIVED WORK
# Filter out segments with less than xx genes above LOQ
# xx is a percentage of target genes out of the total available (18677) and is defined on line 42
xl_metadata<- annot_data.1
xl_counts <-  count_data.1
xl_annot <- xl_metadata %>% dplyr::select(SegmentDisplayName, "ExpressionFilteringThreshold (Human NGS Whole Transcriptome Atlas RNA_1.0)", SegmentLabel, pAlignedReads)
xl_counts <- xl_counts %>% column_to_rownames("TargetName")
xl_counts <- xl_counts[, match(xl_annot$"SegmentDisplayName", colnames(xl_counts))]

## exclude samples with less than XX% of genes above LOQ
LOQ_sample_thresh <- 3 ## set this threshold to whichever you choose as a percent of genes of total available (18677)
LOQ_sample_n <- round(nrow(xl_counts)/100 * LOQ_sample_thresh)
LOQ_sample_n

counts_LOQ_bin <- t(t(xl_counts) > xl_annot$"ExpressionFilteringThreshold (Human NGS Whole Transcriptome Atlas RNA_1.0)")
counts_LOQ_sum_sample <- unname(apply(counts_LOQ_bin, 2, sum)) # count the number of genes above threshold
counts_LOQ_sum_sample

filtered_samples_1 <- xl_annot$SegmentDisplayName[counts_LOQ_sum_sample < LOQ_sample_n]
filtered_samples_1
xl_annot$SegmentDisplayName
sum(counts_LOQ_sum_sample < 560)

# Update the data frames
counts_LOQ_bin <- counts_LOQ_bin[, !(colnames(counts_LOQ_bin) %in% filtered_samples_1)]
xl_annot <- xl_annot[!(xl_annot$SegmentDisplayName %in% filtered_samples_1), ]
xl_counts <- xl_counts[, !(colnames(xl_counts) %in% filtered_samples_1)]
# ====== Target filtering ======
# exclude genes that are below LOQ in xx% of segments
LOQ_gene_thresh <- 10 ## define your own % threshold here
LOQ_gene_n <- round(ncol(xl_counts)/100 * LOQ_gene_thresh)
counts_LOQ_sum_gene <- unname(apply(counts_LOQ_bin, 1, sum)) # count the number of samples in
# which the gene is above threshold
filtered_genes <- rownames(counts_LOQ_bin)[counts_LOQ_sum_gene < LOQ_gene_n]
print(paste0("Threshold: Genes that are above LOQ in ", LOQ_gene_thresh, "% of samples. ", length(filtered_genes), " genes excluded out of ", nrow(counts_LOQ_bin)))
# Update the data frames
counts_LOQ_bin <- counts_LOQ_bin[!(rownames(counts_LOQ_bin) %in% filtered_genes), ]
xl_counts <- xl_counts[!(rownames(xl_counts) %in% filtered_genes), ]
xl_counts <- tibble::rownames_to_column(xl_counts, "TargetName")
################# End of Filtering steps for segments and targets #################
# ====== NEGPROB ROW FOR IDENTIFICATION ======
negp_row <- count_data.1[count_data.1$TargetName == "NegProbe-WTX",]
negp_row <- as.data.frame(negp_row)
row.names(negp_row) <- negp_row$TargetName
#negp_row <- negp_row[, -which(names(negp_row) == "TargetName")]
# remove the negp_row TME sample that are below LOQ:
columns_to_keep <- setdiff(colnames(negp_row), filtered_samples_1)
negp_row_filtered <- negp_row[, columns_to_keep]
#xl_counts <- as.matrix(xl_counts)
xl_counts <- rbind(xl_counts, negp_row_filtered)
tail(xl_counts)
# ===== Special Data Adaptation =====
# for tme only, when samples were excluded:
xl_annot <- subset(xl_annot, SegmentDisplayName %in% annot_data.1$SegmentDisplayName)
match_indices <- match(xl_annot$SegmentDisplayName, annot_data.1$SegmentDisplayName)
xl_annot$PatientNum <- annot_data.1$PatientNum[match_indices]
xl_annot$PatientNum <- as.factor(xl_annot$PatientNum)
# end
xl_annot$U6ATAC_RNAscopeClass <- annot_data.1$U6ATAC_RNAscopeClass[match(xl_annot$SegmentDisplayName, annot_data.1$SegmentDisplayName)]
#xl_annot$PatientNum <- annot_data.1$PatientNum[annot_data.1$SegmentDisplayName == xl_annot$SegmentDisplayName]
#xl_annot$PatientNum <- as.factor(xl_annot$PatientNum)
xl_annot
colData <- xl_annot %>%
  dplyr::select(SegmentDisplayName, U6ATAC_RNAscopeClass, SegmentLabel, PatientNum) %>%
  distinct()  # Ensure unique rows if needed
colData$condition <- with(colData, paste(U6ATAC_RNAscopeClass, SegmentLabel, sep="_"))
rownames(colData) <- colData$SegmentDisplayName
print(head(colData))
# change the first row to column name to maintain gene names, and order them
countData <- xl_counts
countData <- as.data.frame(countData)
rownames(countData) <- countData$TargetName
#remove the extra column of names
countData <- countData[,-1 ]
head(countData)
# Sorting stage is doing something weird.
correct_order <- colData$SegmentDisplayName
countData <- countData[, correct_order]
all(colnames(countData) == correct_order)  # This should return TRUE

# makr matrix and round them. 
countData <- as.matrix(countData)
countData <- round(countData)

dds <- DESeqDataSetFromMatrix(countData = countData,
                              colData = colData,
                              design = ~ U6ATAC_RNAscopeClass)
# Running the DESeq2 analysis
dds <- DESeq(dds)
dds_vst<- vst(dds, blind=TRUE)
dds_counts <- counts(dds, normalized=TRUE)
#pca_plot <- plotPCA(dds_vst, intgroup = "condition", returnData = FALSE)
#pca_plot <- plotPCA(dds_vst, intgroup = "PatientNum", returnData = FALSE)
pca_plot <- plotPCA(dds_vst, intgroup = "U6ATAC_RNAscopeClass", returnData = FALSE)
#pca_plot <- plotPCA(dds_counts, intgroup = "U6ATAC_RNAscopeClass", returnData = FALSE)
print(pca_plot)
# slice conditions of interest
dds_tumour<- results(dds, contrast=c("U6ATAC_RNAscopeClass", "high","low"))
#dds_tme <- results(dds, contrast=c("condition", "high_TME", "low_TME"))
#dds_tme<- results(dds, contrast=c("U6ATAC_RNAscopeClass", "high","low"))
dds_tumour[rownames(dds_tumour)=="TP53",]
# =====Onwards with the volcanos =====
#Investigate gene of interest, e.g. TP53
gene_name <- "TP53"
counts <- plotCounts(dds, gene = gene_name, intgroup = "U6ATAC_RNAscopeClass", returnData = TRUE)
counts <- tibble::rownames_to_column(counts, "SegmentDisplayName")
pos <- position_jitter(width =0.1, seed = 1)
# Visualize the result
counts %>%
  ggplot(aes(x = U6ATAC_RNAscopeClass, y = count, fill = U6ATAC_RNAscopeClass)) +
  geom_boxplot(outlier.color = NA) +
  geom_jitter(position = pos) +
  theme_bw() +
  theme(axis.title.x.bottom = element_blank(), legend.position = "none") +
  labs(title = paste("Counts of", gene_name, "by condition, proper"))

#===== Violin plot of genes of interest =====
# Assuming gene_name and counts dataframe are already set as needed
# gene_name <- "TP53"  # Your GOI

# Using the dataframe generated from plotCounts
# counts <- plotCounts(dds, gene = gene_name, intgroup = "U6ATAC_RNAscopeClass", returnData = TRUE)
# Plotting the violin plot
# ggplot(counts, aes(x = U6ATAC_RNAscopeClass, fill = U6ATAC_RNAscopeClass, y = count)) +
#  geom_violin() +
#  geom_jitter(position = position_jitter(width = 0.3, seed = 2)) +  # Adjust jitter position with specified width
#  labs(title = paste("Expression of", gene_name, "by Condition"), y = paste(gene_name, "Expression")) +
#  scale_y_continuous(trans = "log2", name = paste(gene_name, "Expression (log2 scale)")) +
#  scale_fill_manual(values = c("high" = "#1f77b4", "low" = "#ff7f0e")) +  # Customize colors here
#  theme_bw() +
#  theme(axis.title.x = element_blank(), legend.position = "none")

# ===== list of genes of interest =====
dna_repair_genes_all <- c("BRCA1", "BRCA2", "ATM", "CHEK2", "PALB2", "RAD51", "XRCC1",
                          "XRCC2", "XRCC3", "RAD51C", "RAD51D", "RAD52", "RAD54L",
                          "BRIP1", "BARD1", "NBN", "RAD50", "MRE11", "ATR",
                          "FANCA", "FANCB", "FANCC", "FANCD2", "FANCE", "FANCF",
                          "BRCC3", "FANCG", "FANCI", "UBE2T", "FANCL", "FANCM",
                          "PTEN", "MLH1", "MSH2", "MSH3", "MSH6", "PMS1",
                          "PMS2", "POLD1", "POLE", "EXO1", "RAD1", "RAD9A", "HUS1",
                          "RAD17", "TOPBP1", "MDC1")
MIG_list <- read_excel("ResponsiveIntrons.xlsx")
old_mig <- read_excel("old_MIGs.xlsx",sheet=1)
mig_genes <- old_mig$`All MIGs`
e2f_genes <- c("E2F1", "E2F2", "E2F3", "E2F4", "E2F5", "E2F6", "E2F7", "E2F8") 
DNA_r_144_genes <- read_excel("DNA_repair_genes_144.xlsx")
DNA_r_144_genes <- DNA_r_144_genes$list_of_144_genes

# ===== Convert Data for GSEA analysis ====
results_df_gsea <- results_df %>% rownames_to_column(var="SYMBOL")
ens2symbol <- AnnotationDbi::select(org.Hs.eg.db,
                                    key=results_df_gsea$SYMBOL, 
                                    columns=("ENTREZID"), 
                                    keytype="SYMBOL") 
ens2symbol <- dplyr::as_tibble(ens2symbol)
res <- dplyr::inner_join(ens2symbol, results_df_gsea, by="SYMBOL")
res2 <- res %>% 
  dplyr::select(ENTREZID, log2FoldChange) %>% 
  na.omit() %>% 
  dplyr::distinct() %>% 
  group_by(ENTREZID) %>%  ## this averages the log2fc across duplicate genes
  dplyr::summarize(stat=mean(log2FoldChange)) # stat variable is now equivalent to log2fc

res2 <- as.data.frame(res2)
geneList = res2[,2] # numerical variable, gene counts, log2fc
names(geneList) = as.character(res2[,1]) # character variable for gene names
geneList = sort(geneList, decreasing = TRUE) # organizes genelist descending log2fc
#!!!! check this sorting step 
## THIS IS OUR GENESETS AND PATHWAYS TO ASSESS
library("msigdbr")
H <- msigdbr(species = "Homo sapiens", category = "H") %>% 
  dplyr::select(gs_name, entrez_gene)

# perform GSEA
em2 <- GSEA(geneList, TERM2GENE = H)

# visualise GSEA DOTPLOT
category_num = 50
dot1 <- dotplot(em2, showCategory=category_num, split=".sign", font.size=8, size = NULL, label_format = 90) + 
  ggtitle(paste("top",category_num, " hallmarks of", tag, "high vs low U6atac")) +
  facet_grid(~.sign)
dot1 + 
  theme(plot.title.position = "plot")

# Extract relevant information
category_num <- 50
top_categories <- head(em2$ID, 10)
category_NES <- tail(em2$NES, category_num)
em2$ID
  
# Create a data frame for plotting
plot_data <- data.frame(em2)

# Create the bar plot
category_colors <- c("HALLMARK_E2F_TARGETS" = "GOI",
                     "Other" = "Others",
                     "HALLMARK_DNA_REPAIR" = "GOI"# Default color for other categories
)
print(em2$ID)
# Map category colors to the plot data
plot_data$Color <- ifelse(plot_data$ID == "HALLMARK_E2F_TARGETS" | plot_data$ID == "HALLMARK_DNA_REPAIR",
                          category_colors["HALLMARK_E2F_TARGETS"],
                          category_colors["Other"]
)

# Create the bar plot
bar_plot <- ggplot(plot_data, aes(x = reorder(plot_data$ID, NES), y = NES, fill = Color)) +
  geom_bar(stat = "identity") +  # Define the gradient colors
  coord_flip() +
  labs(title = paste("Top", category_num, "hallmarks of", tag, "high vs low U6atac"),
       x = "Hallmark", y = "NES") +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5))

print(bar_plot)
bar_plot <- ggplot(plot_data, aes(x = reorder(plot_data$ID, NES), y = NES)) +
  geom_bar(stat = "identity", fill = "skyblue") +
  coord_flip() +  # Flip the coordinates to create a horizontal bar plot
  labs(title = paste("Top", category_num, "hallmarks of", tag, "high vs low U6atac"),
       x = "Hallmark", y = "Normalized Enrichment Score") +
  theme_minimal() 
print(bar_plot)

#filename <- paste("top", category_num, "hallmarks of", tag, "high vs low U6atac", ".png", sep = "_")
#ggsave(filename, plot = dot1, device = "png")

## visualise specific pathways 
em2@result[["ID"]]
gseaplot2(em2, geneSetID = 8, title = em2$Description[8]) #Works
temp_gsea <- em2@result
## OVER REPRESENTATION ANALYSIS AND VISUALISATION
data(geneList, package="DOSE")
gene <- names(geneList)
em <- enricher(gene, TERM2GENE=H)
## visualise ORA
upsetplot(em)

#check & filter 
results_df <- as.data.frame(dds_tumour)
rownames(results_df)
results_df$gene <- rownames(results_df)
results_df$negLog10Pvalue <- -log10(results_df$pvalue)
# ===== Additional GOIs =====
e2f_list <- read_excel("E2F targets.xlsx",sheet=2)
e2f_c <- e2f_list$C1_cell_cycle
e2f_checkpoint <- e2f_list$checkpoint
ef2_synthesis <-e2f_list$`ii DNA synthesis`
e2f_differentiation <- e2f_list$`(viii) Differentiation`
e2f_BPAX1 <- e2f_list$`(vii) Development BAPX1`
e2f_apop <- e2f_list$`(vi) Apoptosis`
e2f_DNA_repair <- e2f_list$`(v) DNA damage rep`
e2f_dna_syn <- e2f_list$`ii DNA synthesis`
# ------

# ===== Visualize Results focusing on DNA repair genes ====
# Define the key variables
results_df <- dds_tumour
TOI <- DNA_r_144_genes
NOI <- 'DNA_repair 144'
NOI_SIG <- '144 DNA repair Significant'
NOI_UNSIG <- '144 DNA repair NOT Significant'

# Ensure column is properly created
results_df$highlight <- ifelse(rownames(results_df) %in% TOI, NOI, "Other")
# Subset for DNA results based on highlight
dna_results <- results_df[results_df$highlight == NOI, ]
print(dna_results)  # Verify the subset
results_df$color_group <- ifelse(results_df$highlight == NOI & results_df$pvalue < 0.05, 
                                 NOI_SIG, 
                                 ifelse(results_df$highlight == NOI, NOI_UNSIG, "Other"))
# Set factor levels explicitly for color_group
results_df$color_group <- factor(results_df$color_group, levels = c("Other", NOI_UNSIG, NOI_SIG))
# Check unique values in color_group
print(unique(results_df$color_group))

# Check the first few rows to confirm
head(results_df)
print(results_df["SMC1A", ])
results_df <- na.omit(results_df)  # Remove NA values to prevent issues in plotting
sum(results_df$color_group == NOI_UNSIG)
sum(results_df$color_group == NOI_SIG)
sum(results_df$color_group == NOI_SIG & results_df$log2FoldChange > 0)
sum(results_df$color_group == NOI_SIG & results_df$log2FoldChange < 0)
# save these
paper_144_sig_up <- results_df[results_df$color_group == NOI_SIG & results_df$log2FoldChange > 0, ]
# Specify the path and file name for your Excel file
path <- "filtered_results_144_sig_up.xlsx"
paper_144_sig_down <- results_df[results_df$color_group == NOI_SIG & results_df$log2FoldChange < 0, ]
path <- "filtered_results_144_sig_down.xlsx"
library(openxlsx)
# Write the data to an Excel file
write.xlsx(paper_144_sig_up, file = path)
write.xlsx(paper_144_sig_down, file = path)

# Setting the factor levels explicitly to match the order in scale_color_manual
results_df$color_group <- factor(results_df$color_group, levels = c("Other", "144 DNA repair NOT Significant", "144 DNA repair Significant"))
# Generate the volcano plot
volcano_plot1 <- ggplot(results_df, aes(x = log2FoldChange, y = negLog10Pvalue)) +
  # Plot "Other" points first
  geom_point(data = subset(results_df, color_group == "Other"),
             aes(color = color_group), alpha = 0.5, size = 3) +
  # Plot significant points next
  geom_point(data = subset(results_df, color_group == "144 DNA repair Significant"),
             aes(color = color_group), alpha = 1, size = 4) +
  # Plot non-significant points last
  geom_point(data = subset(results_df, color_group == "144 DNA repair NOT Significant"),
             aes(color = color_group), alpha = 0.5, size = 3) +
  labs(title = "Tumour. high vs low U6atac", x = "Log2 Fold Change", y = "-Log10 p-value", color = "Gene Group") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "blue") +
  geom_vline(xintercept = 0, linetype = "dotted") +
  annotate("text", x = -max(results_df$log2FoldChange) * 0.1, y = max(results_df$negLog10Pvalue)*1.1, label = "High U6atac", hjust = -3, vjust = 4) +
  annotate("text", x = max(results_df$log2FoldChange) * 0.1, y = max(results_df$negLog10Pvalue)*1.1, label = "Low U6atac", hjust = 3, vjust = 4) +
  theme(axis.line = element_line(color='white'), panel.background = element_rect(fill = "white"),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.border = element_blank()) +
  scale_color_manual(values = c("Other" = "grey", 
                                "144 DNA repair Significant" = "red", 
                                "144 DNA repair NOT Significant" = "#b48080")) +
  geom_text_repel(data = subset(results_df, color_group == "144 DNA repair Significant"),
                  aes(label = rownames(subset(results_df, color_group == "144 DNA repair Significant"))),
                  size = 3, color = "black", box.padding = 0.35, point.padding = 0.5,
                  min.segment.length = 0, max.overlaps = Inf)

#ggsave("volcano_plot.png", plot = volcano_plot1, width = 10, height = 8, units = "in")
print(volcano_plot1)
#ggsave("low U6, tumour vs TME, volcano_plot, DNA repair genes.pdf", volcano_plot, width = 10, height = 8,units = "in")
dev.off()
# ===== Alternative Way to Plot the same data ====
#volcano_plot2<- ggplot(results_df, aes(x = log2FoldChange, y = negLog10Pvalue)) +
#  geom_point(data = subset(results_df, color_group == "Other"),
#             aes(color = color_group), alpha = 0.5, size = 3) +
#  geom_point(data = subset(results_df, color_group %in% c("DNA repair Significant",
#                                                          "DNA repair Not Significant")),
#             aes(color = color_group), size = 3.5) +
#  scale_color_manual(values = c("Other" = "grey", 
#                                "DNA repair Significant" = "red", 
#                                "DNA repair Not Significant" = "#b48080")) +
#  labs(title = "Tumour. high vs low",
#       x = "Log2 Fold Change", y = "-Log10 p-value",
#       color = "Gene Group") +
#  theme_minimal() +
#  geom_text_repel(data = subset(results_df, color_group %in% c("DNA repair Significant", "DNA repair Not Significant")),
#                  aes(label = rownames(subset(results_df, color_group %in% c("DNA repair Significant", "DNA repair Not Significant")))),
#                  size = 3, color = "black", box.padding = 0.35, point.padding = 0.5,
#                  min.segment.length = 0, max.overlaps = Inf)
#print(volcano_plot2)
#plot_tumour <- grid.arrange(volcano_plot1, volcano_plot2, nrow = 1)
#ggsave("Tumour.pdf", plot_tumour, width = 20, height = 8,units = "in")

# =====Checking E2Fs=====
all_e2f <- read_excel("E2F targets.xlsx",sheet=1)
all_e2f <- all_e2f$GeneName
results_df <- dds_tumour
results_df <- as.data.frame(results_df)
print(rownames(results_df)[rownames(results_df) %in% all_e2f])
results_df$gene <- rownames(results_df)
results_df$negLog10Pvalue <- -log10(results_df$pvalue)
# Adding the highlight column correctly
results_df$highlight <- ifelse(rownames(results_df) %in% all_e2f,
                               "E2F target", "Other")
dna_results <- results_df[results_df$highlight == "E2F target", ]
print(dna_results)

results_df$color_group <- ifelse(results_df$highlight == "E2F target" & results_df$pvalue < 0.05, 
                                 "E2F target Significant", 
                                 ifelse(results_df$highlight == "E2F target",
                                        "E2F target Not Significant", "Other"))


# Check the first few rows to confirm 'highlight' column exists
head(results_df)
results_df["EZH2", ]
# Same plotting method used here. 
# Create the basic volcano plot with adjusted text labels
volcano_plot1 <- ggplot(results_df, aes(x = log2FoldChange, y = negLog10Pvalue)) +
  geom_point(data = subset(results_df, color_group == "Other"),
             aes(color = color_group), alpha = 0.5, size = 3) +
  geom_point(data = subset(results_df, color_group %in% c("E2F target Significant",
                                                          "E2F target Not Significant")),
             aes(color = color_group), size = 3.5) +
  scale_color_manual(values = c("Other" = "grey", 
                                "E2F target Significant" = "red", 
                                "E2F target Not Significant" = "#b48080")) +
  labs(title = "Tumour. high vs low U6atac",
       x = "Log2 Fold Change", y = "-Log10 p-value",
       color = "Gene Group") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "blue") +
  geom_vline(xintercept = 0, linetype = "dotted") +
  annotate("text", x = -max(results_df$log2FoldChange) * 0.1, y = max(results_df$negLog10Pvalue)*1.1, label = "High U6atac", hjust = -3, vjust = 4) +
  annotate("text", x = max(results_df$log2FoldChange) * 0.1, y = max(results_df$negLog10Pvalue)*1.1, label = "Low U6atac", hjust = 3, vjust = 4) +
  theme(axis.line = element_line(color='white'), panel.background = element_rect(fill = "white"),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        panel.border = element_blank()) +
  geom_text_repel(data = subset(results_df, color_group %in% "E2F target Significant"),
                  aes(label = rownames(subset(results_df, color_group %in% "E2F target Significant"))),
                  size = 3, color = "black", box.padding = 0.35, point.padding = 0.5,
                  min.segment.length = 0, max.overlaps = Inf)
print(volcano_plot1)

# =====BPAX1 Check=====
results_df <- dds_tumour
results_df <- as.data.frame(results_df)
e2f_BPAX1 <- na.omit(e2f_BPAX1)
print(rownames(results_df)[rownames(results_df) %in% e2f_BPAX1])
results_df$gene <- rownames(results_df)
results_df$negLog10Pvalue <- -log10(results_df$pvalue)
# Adding the highlight column correctly
results_df$highlight <- ifelse(rownames(results_df) %in% e2f_BPAX1,
                               "E2F BPAX1", "Other")
dna_results <- results_df[results_df$highlight == "E2F BPAX1", ]
print(dna_results)
results_df$color_group <- ifelse(results_df$highlight == "E2F BPAX1" & results_df$pvalue < 0.05, 
                                 "E2F BPAX1 Significant", 
                                 ifelse(results_df$highlight == "E2F BPAX1",
                                        "E2F BPAX1 Not Significant", "Other"))

# Check Integrity
head(results_df)
results_df["TP73", ]
# ===== Plotting BPAX1 =====
# Create the basic volcano plot with adjusted text labels
volcano_plot1 <- ggplot(results_df, aes(x = log2FoldChange, y = negLog10Pvalue)) +
  geom_point(data = subset(results_df, color_group == "Other"),
             aes(color = color_group), alpha = 0.5, size = 3) +
  geom_point(data = subset(results_df, color_group %in% c("E2F BPAX1 Significant",
                                                          "E2F BPAX1 Not Significant")),
             aes(color = color_group), size = 3.5) +
  scale_color_manual(values = c("Other" = "grey", 
                                "E2F BPAX1 Significant" = "red", 
                                "E2F BPAX1 Not Significant" = "#b48080")) +
  labs(title = "Tumour. high vs low U6atac",
       x = "Log2 Fold Change", y = "-Log10 p-value",
       color = "Gene Group") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "blue") +
  geom_vline(xintercept = 0, linetype = "dotted") +
  annotate("text", x = -max(results_df$log2FoldChange) * 0.1, y = max(results_df$negLog10Pvalue)*1.1, label = "High U6atac", hjust = -3, vjust = 3) +
  annotate("text", x = max(results_df$log2FoldChange) * 0.1, y = max(results_df$negLog10Pvalue)*1.1, label = "Low U6atac", hjust = 3, vjust = 3) +
  theme(axis.line = element_line(color='white'), panel.background = element_rect(fill = "white"),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        panel.border = element_blank()) +
  geom_text_repel(data = subset(results_df, color_group %in% "E2F BPAX1 Significant"),
                  aes(label = rownames(subset(results_df, color_group %in% "E2F BPAX1 Significant"))),
                  size = 3, color = "black", box.padding = 0.35, point.padding = 0.5,
                  min.segment.length = 0, max.overlaps = Inf)
print(volcano_plot1)

# ===== Checking Extended DNA repair GOIs =====
results_df <- dds_tumour
results_df <- dds_tme
results_df <- as.data.frame(results_df)
print(rownames(results_df)[rownames(results_df) %in% dna_repair_genes_all])
results_df$gene <- rownames(results_df)
results_df$negLog10Pvalue <- -log10(results_df$pvalue)
# Adding the highlight column correctly
results_df$highlight <- ifelse(rownames(results_df) %in% dna_repair_genes_all,
                               "DNA repair", "Other")
dna_results <- results_df[results_df$highlight == "DNA repair", ]
print(dna_results)
results_df$color_group <- ifelse(results_df$highlight == "DNA repair" & results_df$pvalue < 0.05, 
                                 "DNA repair Significant", 
                                 ifelse(results_df$highlight == "DNA repair",
                                        "DNA repair Not Significant", "Other"))
head(results_df)
results_df["ATM", ]

volcano_plot1 <- ggplot(results_df, aes(x = log2FoldChange, y = negLog10Pvalue)) +
  geom_point(data = subset(results_df, color_group == "Other"),
             aes(color = color_group), alpha = 0.5, size = 3) +
  geom_point(data = subset(results_df, color_group %in% c("DNA repair Significant",
                                                          "DNA repair Not Significant")),
             aes(color = color_group), size = 3.5) +
  scale_color_manual(values = c("Other" = "grey", 
                                "DNA repair Significant" = "red", 
                                "DNA repair Not Significant" = "#b48080")) +
  labs(title = "TME. high vs low U6atac, DNA repair genes, updated",
       x = "Log2 Fold Change", y = "-Log10 p-value",
       color = "Gene Group") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "blue") +
  geom_vline(xintercept = 0, linetype = "dotted") +
  annotate("text", x = -max(results_df$log2FoldChange) * 0.1, y = max(results_df$negLog10Pvalue)*1.1, label = "High U6atac", hjust = -3, vjust = 3) +
  annotate("text", x = max(results_df$log2FoldChange) * 0.1, y = max(results_df$negLog10Pvalue)*1.1, label = "Low U6atac", hjust = 3, vjust = 3) +
  theme(axis.line = element_line(color='white'), panel.background = element_rect(fill = "white"),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        panel.border = element_blank()) +
  geom_text_repel(data = subset(results_df, color_group %in% "DNA repair Significant"),
                  aes(label = rownames(subset(results_df, color_group %in% "DNA repair Significant"))),
                  size = 3, color = "black", box.padding = 0.35, point.padding = 0.5,
                  min.segment.length = 0, max.overlaps = Inf)
print(volcano_plot1)
sum(results_df$color_group == "DNA repair Significant")
sum(results_df$color_group == "DNA repair Significant" & results_df$log2FoldChange > 0)
sum(results_df$color_group == "DNA repair Significant" & results_df$log2FoldChange < 0)

# =====Checking all E2F targets =====
all_e2f <- read_excel("E2F targets.xlsx",sheet=1)
all_e2f <- all_e2f$GeneName
results_df <- dds_tumour
results_df <- as.data.frame(results_df)
print(rownames(results_df)[rownames(results_df) %in% all_e2f])
results_df$gene <- rownames(results_df)
results_df$negLog10Pvalue <- -log10(results_df$pvalue)
# Adding the highlight column correctly
results_df$highlight <- ifelse(rownames(results_df) %in% all_e2f,
                               "E2F target", "Other")
dna_results <- results_df[results_df$highlight == "E2F target", ]
print(dna_results)

results_df$color_group <- ifelse(results_df$highlight == "E2F target" & results_df$pvalue < 0.05, 
                                 "E2F target Significant", 
                                 ifelse(results_df$highlight == "E2F target",
                                        "E2F target Not Significant", "Other"))


# Check the first few rows to confirm 'highlight' column exists
head(results_df)
results_df["EZH2", ]
# Create the basic volcano plot with adjusted text labels
volcano_plot1 <- ggplot(results_df, aes(x = log2FoldChange, y = negLog10Pvalue)) +
  geom_point(data = subset(results_df, color_group == "Other"),
             aes(color = color_group), alpha = 0.5, size = 3) +
  geom_point(data = subset(results_df, color_group %in% c("E2F target Significant",
                                                          "E2F target Not Significant")),
             aes(color = color_group), size = 3.5) +
  scale_color_manual(values = c("Other" = "grey", 
                                "E2F target Significant" = "red", 
                                "E2F target Not Significant" = "#b48080")) +
  labs(title = "Tumour. high vs low U6atac",
       x = "Log2 Fold Change", y = "-Log10 p-value",
       color = "Gene Group") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "blue") +
  geom_vline(xintercept = 0, linetype = "dotted") +
  annotate("text", x = -max(results_df$log2FoldChange) * 0.1, y = max(results_df$negLog10Pvalue)*1.1, label = "High U6atac", hjust = -3, vjust = 4) +
  annotate("text", x = max(results_df$log2FoldChange) * 0.1, y = max(results_df$negLog10Pvalue)*1.1, label = "Low U6atac", hjust = 3, vjust = 4) +
  theme(axis.line = element_line(color='white'), panel.background = element_rect(fill = "white"),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        panel.border = element_blank()) +
  geom_text_repel(data = subset(results_df, color_group %in% "E2F target Significant"),
                  aes(label = rownames(subset(results_df, color_group %in% "E2F target Significant"))),
                  size = 3, color = "black", box.padding = 0.35, point.padding = 0.5,
                  min.segment.length = 0, max.overlaps = Inf)
print(volcano_plot1)

# =====Checking All Minor Intron Genes =====
results_dds <- dds_tumour
results_df <- as.data.frame(results_dds)
results_df$log2FoldChange <- results_df$log2FoldChange
results_df$negLog10Pvalue <- -log10(results_df$pvalue)

results_df$highlight <- ifelse(rownames(results_df) %in% mig_genes,
                               "MIGs", "Other")
MIG_results <- results_df[results_df$highlight == "MIGs", ]
write.xlsx(as.data.frame(MIG_results), file = "minor_introns_results.xlsx")
# Add gene names as a column in your dataframe
MIG_results$Gene <- rownames(MIG_results)
MIG_results
# Write the dataframe to an Excel file
write.xlsx(as.data.frame(MIG_results), file = "tumour_minor_introns_results.xlsx")
print(MIG_results)
# head(results_df)

#Sanity Check
results_df["STX10", ]
# Create the basic volcano plot with adjusted text labels
volcano_plot1 <- ggplot(results_df, aes(x = log2FoldChange, y = negLog10Pvalue)) +
  geom_point(data = subset(results_df, color_group == "Other"),
             aes(color = color_group), alpha = 0.5, size = 3) +
  geom_point(data = subset(results_df, color_group %in% c("MIGs Significant",
                                                          "MIGs Not Significant")),
             aes(color = color_group), size = 3.5) +
  scale_color_manual(values = c("Other" = "grey", 
                                "MIGs Significant" = "red", 
                                "MIGs Not Significant" = "#b48080")) +
  labs(title = "Tumour high vs low",
       x = "Log2 Fold Change", y = "-Log10 p-value",
       color = "Gene Group") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "blue") +
  theme_minimal()
#geom_text_repel(data = subset(results_df, color_group %in% "MIGs Significant"),
# aes(label = rownames(subset(results_df, color_group %in% "MIGs Significant"))),
# size = 3, color = "black", box.padding = 0.35, point.padding = 0.5,
# min.segment.length = 0, max.overlaps = Inf)
print(volcano_plot1)
MIG_up_tumour <- results_df %>%
  filter(color_group == "MIGs Significant" & log2FoldChange > 0)

MIG_up_tumour$Gene <- rownames(MIG_up_tumour)
write.xlsx(as.data.frame(MIG_up_tumour), file = "tumour_Upregulated_MIGs_updated_U6.xlsx")

MIG_down_tumour <- results_df %>%
  filter(color_group == "MIGs Significant" & log2FoldChange < 0)
MIG_down_tumour$Gene <- rownames(MIG_down_tumour)
write.xlsx(as.data.frame(MIG_down_tumour), file = "tumour_Dowunregulated_MIGs_U6_updated.xlsx")

results_df %>%
  filter(color_group == "MIGs Significant" & log2FoldChange > 0) %>%
  nrow()

results_df %>%
  filter(color_group == "MIGs Significant" & log2FoldChange < 0) %>%
  nrow()
sessioninfo::session_info()
